torch/_inductor/autotune_process.py

```diff
@@ -45,29 +50,30 @@ class TuningProcess:
     requests and return results.
     """
 
+    device: Optional[int] = None
     process: Optional[BaseProcess] = None
-    request_queue: Optional["Queue[Any]"] = None
-    response_queue: Optional["Queue[Any]"] = None
+    request_queue: Optional[Queue[Any]] = None
+    response_queue: Optional[Queue[Any]] = None
 
     @staticmethod
     def process_main(
-        request_queue: "Queue[Any]",
-        response_queue: "Queue[Any]",
+        device: Optional[int],
+        request_queue: Queue[Any],
+        response_queue: Queue[Any],
     ) -> None:
         """
         Entry point for the child process.
         """
-        log.debug("Entering TuningProcess child main")
+        log.debug("Entering TuningProcess child main: %s", device)
+        if device is not None:
+            os.environ[CUDA_VISIBLE_DEVICES] = str(device)
```

Comment:
Interesting that this works. I guess, it means that `pytorch` is able to picks up the device value automatically from the newly update `os.environ["CUDA_VISIBLE_DEVICES"]`?

Reply:
@aakhundov Hmmmmm, good catch. It does not work as I expected. I saw speedups comparable to the first implementation, so blindly assumed it was due to using multiple GPUs. But I guess the use of multiple subprocesses (even if using the same GPU) provides perf benefits? I found `nvidia-smi pmon` and it's reporting all my parallel subprocesses on the same device. The reason for organizing it this way is the multiprocessing API doesn't seem to allow passing a new environment before spawning the new procees. Lemme dig some more.

Reply:
the benchmark result will be very much distorted in this case. We have to make sure different subprocess are running on different devices.

Reply:
I wonder if anyone has any good idea for testing this

Reply:
I agree it's incorrect to run more than one profiling on the same device in parallel. So in AITemplate we do [this](https://www.internalfb.com/code/fbsource/[6640f66ac0bda9a844381e433ab03b14a40b362d]/fbcode/aitemplate/AITemplate/python/aitemplate/backend/profiler_runner.py?lines=226-233) to launch a process with a specific `CUDA_VISIBLE_DEVICES` env var (`dev_select_flag` is `"CUDA_VISIBLE_DEVICES"`; `device` is a zero-based int id of the device). And this seems to work: whatever the process does on a GPU ends up on the #`device`. Could we do something similar here?

Reply:
@masnesral I believe, the reason you saw speedups was because the actual GPU work during benchmarking is not that intensive. As I understand, most of the time may be spent in launching and processing. But we can't guarantee that the kernels will not overlap (hence distort the measurements), so still must serialize them on each device.

Reply:
> So in AITemplate we do [this](https://www.internalfb.com/code/fbsource/%5B6640f66ac0bda9a844381e433ab03b14a40b362d%5D/fbcode/aitemplate/AITemplate/python/aitemplate/backend/profiler_runner.py?lines=226-233)

Yes, setting that env var works as expected, generally speaking. The problem in this current implementation is that setting the env var _after_ the sub-process has already started does not have the desired affect. The previous impl that used a Popen (basically the same approach as the AIT impl) worked correctly in this regard. AFAICT, if we stick with the multiprocessing lib (which has other advantages) the only way to manipulate the env is to do it in the parent process _before_ starting the sub-process. I'll put up an RFC soon.